runtime/sema/check_casting_expression.go

```diff
@@ -412,3 +421,204 @@ func FailableCastCanSucceed(subType, superType Type) bool {
 
 	return true
 }
+
+// IsCastRedundant checks whether a simple cast is redundant.
+// Checks for two cases:
+//    - Case I: Contextually expected type is same as casted type (target type).
+//    - Case II: Expression is self typed, and is same as the casted type (target type).
+func IsCastRedundant(expr ast.Expression, exprInferredType, targetType, expectedType Type) bool {
+	if expectedType != nil && expectedType.Equal(targetType) {
+		return true
+	}
+
+	checkCastVisitor := &CheckCastVisitor{}
+
+	return checkCastVisitor.isCastRedundant(expr, exprInferredType, targetType)
+}
+
+type CheckCastVisitor struct {
+	exprInferredType Type
+	targetType       Type
+}
+
+var _ ast.ExpressionVisitor = &CheckCastVisitor{}
+
+func (d *CheckCastVisitor) isCastRedundant(expr ast.Expression, exprInferredType, targetType Type) bool {
+	prevInferredType := d.exprInferredType
+	prevTargetType := d.targetType
+
+	defer func() {
+		d.exprInferredType = prevInferredType
+		d.targetType = prevTargetType
+	}()
+
+	d.exprInferredType = exprInferredType
+	d.targetType = targetType
+
+	result := expr.AcceptExp(d)
+	return result.(bool)
+}
+
+func (d *CheckCastVisitor) VisitBoolExpression(_ *ast.BoolExpression) ast.Repr {
+	return d.isTypeRedundant(BoolType, d.targetType)
+}
+
+func (d *CheckCastVisitor) VisitNilExpression(_ *ast.NilExpression) ast.Repr {
+	return d.isTypeRedundant(TypeOfNil, d.targetType)
```

Comment:
Maybe `TypeOfNil` should be renamed to `NilType` to be consistent in the naming (cf. `BoolType`). It wasn't introduced in this PR, but would be nice to clean it up 🙏